project_setup/branch-protection.md

```diff
@@ -0,0 +1,36 @@
+# Branch Protection #
+
+We require branch protection to be enabled on all cisagov repositories after
+creation and initial population.
+
+## Setting up a rule to protect the `develop` branch ##
+
+To enable branch protection, you must have permission to access the Settings
+menu.
+
+In `Settings`, go to the `Branches` entry and create a rule with the following:
+
+- Branch name pattern: `develop`
+  - Note: Our primary branches are named `develop`, as a rule.
+  - Verify that adding this name pattern then shows that it
+`Applies to 1 branch: develop`.
+
+- Protect matching branches
+  - [X] Require pull request reviews before merging
+    - Required approving reviews: `2`
+    - [ ] Dismiss stale pull request approvals when new commits are pushed
+    - [X] Require review from Code Owners
+    - [ ] Restrict who can dismiss pull request reviews
+  - [X] Require status checks to pass before merging
+  - [X] Require branches to be up to date before merging
+    - TBD note about status checks
```

Comment:
Reminder that this needs to be fleshed out.

Reply:
How often do we un-check a status check as required? In general, they should all be checked/required, afaik, but do we have any common exceptions that y'all know of?

Reply:
It's pretty rare that we uncheck a status as required.  These days, I think it only happens on older repos that have older-style checks still enabled/required (i.e. ones that we have made optional on newer repos like LGTM or Coveralls).  Another time we might disable a check is if the service that performs the check is down or otherwise malfunctioning.

Reply:
It's pretty rare, as @dav3r said.  There are some older repos that have some erroneous LGTM javascript checks that always fail, and we uncheck those.  (They can only be turned off by sending an email to LGTM and asking them to manually sort it out.)  New repos should not have this issue.  

On very rare occasions we manually disable checks to merge a PR in order to fix something critical that is broken in production, if the checks are failing due to the GitHub Actions service, say, being down.

Both cases are the exception rather than the rule, though, so I don't think they need to be mentioned.

Reply:
An item that probably _should_ be mentioned is that the list of status checks doesn't fully populate until a PR has been created.  This is a minor problem when one creates the repo by pushing directly to `develop` instead of pushing to another branch - the checkboxes in that list can't be checked until the first PR is created.

This problem could be alleviated if we did this (which I only very rarely do myself right now):
1. Create the repo from a skeleton via our `skeleton` tool
2. Push the repo to `develop` _before making any changes from the skeleton code_
3. Create a new branch called `first_commits` or something like that and customize the skeleton, add code, etc. until the new repo is ready
4. Create a PR to merge `first commits` into `develop`

Another advantage of this approach is that it is easy to see in that first PR where the repo creator modified the skeleton code when setting up the new repo.

I don't think it is much extra work to start doing things this way, and if we document the process and scold each other it will soon become habit.  Thoughts, @cisagov/team-ois?

Reply:
I like @jsf9k's solution a lot- it gets a 👍 from me.

Reply:
I think the initial PR should still be reviewed- it's nice for visibility to the rest of the team, plus good practice in general in case there's something wonky in there.

Reply:
I agree with @dav3r 1000%.  I don't like the way we just push unreviewed code to `develop` and run with it when starting a new repo.

Reply:
Added in https://github.com/cisagov/development-guide/pull/44/commits/a645ed5182a376e39efe9cd9e3b4f0c96f33b4c7